src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java

```diff
@@ -81,12 +48,55 @@
 import io.fabric8.kubernetes.client.dsl.PrettyLoggable;
 import jenkins.model.Jenkins;
 import jenkins.model.JenkinsLocationConfiguration;
+import org.csanchez.jenkins.plugins.kubernetes.pipeline.PodTemplateStepExecution;
+import org.csanchez.jenkins.plugins.kubernetes.volumes.PodVolume;
+import org.jenkinsci.plugins.durabletask.executors.OnceRetentionStrategy;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.DataBoundSetter;
+import org.kohsuke.stapler.QueryParameter;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import java.io.IOException;
+import java.net.SocketTimeoutException;
+import java.net.URL;
+import java.nio.file.Paths;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.UnrecoverableKeyException;
+import java.security.cert.CertificateEncodingException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import static java.lang.String.format;
+import static org.apache.commons.lang.StringUtils.isBlank;
+import static org.apache.commons.lang.StringUtils.isNotBlank;
+import static org.csanchez.jenkins.plugins.kubernetes.PodEnvVar.Names.HOME;
+import static org.csanchez.jenkins.plugins.kubernetes.PodEnvVar.Names.JENKINS_JNLP_URL;
+import static org.csanchez.jenkins.plugins.kubernetes.PodEnvVar.Names.JENKINS_LOCATION_URL;
+import static org.csanchez.jenkins.plugins.kubernetes.PodEnvVar.Names.JENKINS_NAME;
+import static org.csanchez.jenkins.plugins.kubernetes.PodEnvVar.Names.JENKINS_SECRET;
+import static org.csanchez.jenkins.plugins.kubernetes.PodEnvVar.Names.JENKINS_TUNNEL;
+import static org.csanchez.jenkins.plugins.kubernetes.PodEnvVar.Names.JENKINS_URL;
+import static org.csanchez.jenkins.plugins.kubernetes.PodTemplateUtils.substituteEnv;
 
 /**
  * Kubernetes cloud provider.
- * 
+ *
  * Starts slaves in a Kubernetes cluster using defined Docker templates for each label.
- * 
+ *
  * @author Carlos Sanchez carlos@apache.org
  */
 public class KubernetesCloud extends Cloud {
```

Comment:
The whole class is daunting, it is larger than 1000 LOC and it does not follow step-down rule. 
Methods could be rearranged so that it would be possible to read it from top to bottom,
and provisioning callbacks could be potentially moved to a separate top-level classes,
especially since they accept the `KubernetesCloud` in a constructor. 
But this is a large refactoring task for a separate PR.